src/vm/eventpipe.cpp

```diff
@@ -0,0 +1,234 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "eventpipe.h"
+#include "eventpipejsonfile.h"
+#include "sampleprofiler.h"
+
+CrstStatic EventPipe::s_initCrst;
+bool EventPipe::s_tracingInitialized = false;
+bool EventPipe::s_tracingEnabled = false;
+EventPipeJsonFile* EventPipe::s_pJsonFile = NULL;
+
+void EventPipe::Initialize()
+{
+    STANDARD_VM_CONTRACT;
+
+    s_tracingInitialized = s_initCrst.InitNoThrow(
+        CrstEventPipe,
+        (CrstFlags)(CRST_TAKEN_DURING_SHUTDOWN));
+}
+
+void EventPipe::EnableOnStartup()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    // Test COMPLUS variable to enable tracing at start-up.
+    if(CLRConfig::GetConfigValue(CLRConfig::INTERNAL_PerformanceTracing) != 0)
+    {
+        Enable();
+    }
+}
+
+void EventPipe::Shutdown()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    Disable();
+}
+
+void EventPipe::Enable()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(!s_tracingInitialized)
+    {
+        return;
+    }
+
+    // Take the lock and enable tracing.
+    CrstHolder _crst(&s_initCrst);
+    s_tracingEnabled = true;
+    if(CLRConfig::GetConfigValue(CLRConfig::INTERNAL_PerformanceTracing) == 2)
+    {
+        // File placed in current working directory.
+        SString outputFilePath;
+        outputFilePath.Printf("Process-%d.PerfView.json", GetCurrentProcessId());
+        s_pJsonFile = new EventPipeJsonFile(outputFilePath);
+    }
+
+    SampleProfiler::Enable();
+}
+
+void EventPipe::Disable()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    CrstHolder _crst(&s_initCrst);
+    s_tracingEnabled = false;
+    SampleProfiler::Disable();
+
+    if(s_pJsonFile != NULL)
+    {
+        delete(s_pJsonFile);
+        s_pJsonFile = NULL;
+    }
+}
+
+bool EventPipe::EventEnabled(GUID& providerID, INT64 keyword)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    // TODO: Implement filtering.
+    return false;
+}
+
+void EventPipe::WriteEvent(GUID& providerID, INT64 eventID, BYTE *pData, size_t length, bool sampleStack)
```

Comment:
I am thinking that we may wish to have abstractions for both providers and events (thus you pass a Provider&  or frankly just a Event& (which points at the provider).    I really don't like GUIDs and I would like to hide them in most code.